instantout/reservation/manager.go

```diff
@@ -0,0 +1,259 @@
+	if err != nil {
+		return err
+	}
+
+	reservationResChan := make(
+		chan *reservationrpc.ServerReservationNotification,
+	)
+
+	err = m.RegisterReservationNotifications(runCtx, reservationResChan)
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case height := <-newBlockChan:
+			log.Debugf("Received block %v", height)
+			currentHeight = height
+
+		case reservationRes := <-reservationResChan:
+			log.Debugf("Received reservation %x",
+				reservationRes.ReservationId)
+			err := m.newReservation(
+				runCtx, uint32(currentHeight), reservationRes,
+			)
+			if err != nil {
+				return err
+			}
+
+		case err := <-newBlockErrChan:
+			return err
+
+		case <-runCtx.Done():
+			log.Debugf("Stopping reservation manager")
+			return nil
+		}
+	}
+}
+
+// newReservation creates a new reservation from the reservation request.
+func (m *Manager) newReservation(ctx context.Context, currentHeight uint32,
+	req *reservationrpc.ServerReservationNotification) error {
+
+	var reservationID ID
+	err := reservationID.FromByteSlice(
+		req.ReservationId,
+	)
+	if err != nil {
+		return err
+	}
+
+	serverKey, err := btcec.ParsePubKey(req.ServerKey)
+	if err != nil {
+		return err
+	}
+
+	// Create the reservation state machine. We need to pass in the runCtx
+	// of the reservation manager so that the state machine will keep on
+	// running even if the grpc conte
+	reservationFSM := NewFSM(
+		ctx, m.cfg,
+	)
+
+	initContext := &InitReservationContext{
+		reservationID: reservationID,
+		serverPubkey:  serverKey,
+		value:         btcutil.Amount(req.Value),
+		expiry:        req.Expiry,
+		heightHint:    currentHeight,
+	}
+
+	// Send the init event to the state machine.
+	err = reservationFSM.SendEvent(OnServerRequest, initContext)
+	if err != nil {
+		return err
+	}
+
+	// We'll now wait for the reservation to be in the state where it is
+	// waiting to be confirmed.
+	err = reservationFSM.DefaultObserver.WaitForState(
+		ctx, time.Minute, WaitForConfirmation,
+		fsm.WithWaitForStateOption(time.Second),
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RegisterReservationNotifications registers a new reservation notification
+// stream.
+func (m *Manager) RegisterReservationNotifications(
+	ctx context.Context, reservationChan chan *reservationrpc.
+		ServerReservationNotification) error {
+
+	// In order to create a valid lsat we first are going to call
+	// the FetchL402 method.
+	_, err := m.cfg.ReservationClient.FetchL402(
+		ctx, &reservationrpc.FetchL402Request{},
+	)
+	if err != nil {
+		return err
+	}
+
+	// We'll now subscribe to the reservation notifications.
+	reservationStream, err := m.cfg.ReservationClient.
+		ReservationNotificationStream(
+			ctx, &reservationrpc.ReservationNotificationRequest{},
+		)
+	if err != nil {
+		return err
+	}
+
+	// We'll now start a goroutine that will forward all the reservation
+	// notifications to the reservationChan.
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
```

Comment:
I think we don't need this check here, since the stream will fail if the context is canceled anyway. And also we below check in the reconnect loop too.